### DISCUSSION TITLE
请问自动代理怎么用的??

### DISCUSSION THREAD

**Discussion by k79e** [SeaHOH/GotoX#124] [问答/咨询]
我看到这个能直接下载路由表. 可是下载完了怎么用哇!!
是不是这个功能还没做完整?

**Comment by SeaHOH** [comment]
简单的讲，就是根据内置逻辑和用户规则来自动选择链接路径，路由表和域名列表都是这个逻辑的部分判断依据。具体配置方法请阅读配置文件注释。

**Comment by k79e** [comment]
我粗略看了下 好像没看见那个讲路由表的哇!!
天哪是不是我看走眼了
代理直连   | direct    |  是  | 无／IP／主配置 IP 列表名称／域名，注1
就这地方我还是不太懂
啥叫名称啊. 那个表有apnic的和17mon的
比如我要用17mon的 我就写17mon就行了??

**Comment by SeaHOH** [comment]
上面讲的内置逻辑的一部分就是：除非在规则中指定，否则绝对不使用 GAE 代理包含于路由表和域名列表的主机。

“主配置 IP 列表名称”是指在主配置的 `iplist` 小节定义的列表的名称，可以自定义，注意事项看注释。
默认自动规则就是一种示例，对照注释看基本就能理解了。

**Comment by k79e** [comment]
啊!!? 怎么又绕回iplist了!!!
请问ip库叫什么名字啊. iplist可以填那个txt文件的名称是吧比如这个directdomains.txt
那ip库不是txt啊 我直接写directip 就行了??

现在我看这个规则是没写的 不匹配的一律走直连.
我怎么让他反着来 走白名单模式呢?

**Comment by SeaHOH** [comment]
这个 `iplist` 就相当于 `Hosts`，作用是类似的。
你在 `iplist` 定义一个列表，其名称可用于自动规则，其 IP 就被自动规则匹配的域名使用。

**Comment by SeaHOH** [comment]
> 我怎么让他反着来 走白名单模式呢?

你所谓的白名单模式是指什么模式呢？任何所谓黑白名单，它都有一个作用范围。

**Comment by k79e** [comment]
比如我弄个表a啊 他都是中国ip 我让他走代理
然后另一个地方排除下a不就行了 就相当于白名单模式了 只有a不走代理 其他全走代理.

**Comment by SeaHOH** [comment]
> 比如我弄个表a啊 他都是中国ip 我让他走代理

上面已经把这个逻辑固化在程序中，就是中国 IP (即匹配路由表) 不允许使用 GAE 代理。
在这之前还有一个逻辑就是，当直连检测到某些故障时会**自动**尝试使用 GAE 代理。

我正在写一个关于自动代理动作说明的 wiki，完成之后你可以看看。

**Comment by k79e** [comment]
原来是自动的啊!!!
现在就缺个白名单模式了 要是其余自动走代理那真是一劳永逸!!
不过那样跟sni可能会冲突
但是如果把sni排除下再给个选项 那样就完美了!!

**Comment by k79e** [comment]
奇怪了这是什么原理
我用v2ray做了跳转 走gotox出去
载入静态资源没问题
但是打开这个页面就有问题. 就是当前页面
![image](https://user-images.githubusercontent.com/6483602/64846003-162ee800-d63e-11e9-985b-6b921ae700d1.png)
啊!!!!! 可能是v2配置有问题要不然就是不能这样用... 发现是所有tls都打不开了...
2019-09-13 16:21:53
果然是v2的问题

**Comment by k79e** [comment]
对了我还有点小小的疑问

就是在我使用绑定域名模式的时候
我想给这个域名指定一些ip使用. 但是我发现这个软件不是最优先读取hosts查询dns的.
配置里面好像也没找到什么办法.
不知道目前能不能那样用呢.

**Comment by SeaHOH** [comment]
> 不知道目前能不能那样用呢.

之前是有打算这样用，但是现在确实不行。

**Comment by SeaHOH** [comment]
9add208 已支持自定义绑定域名的 IP，使用时在自动代理配置的 `forward` 或 `direct` 动作中进行设置。需要注意的是，自定义的 IP 是否能返回正确的证书和代理服务。

**Comment by k79e** [comment]
天哪 太好用了!!!!

**Comment by k79e** [comment]
能否加个测试 就是建立ssl/tls连接的时候.
是传输网站数据的时候 建立连接时候转发一下浏览器端的alpn (不是先建立到gae的那个连接 给gae连接用alpn. 是隧道内用户数据 浏览器发送的这个tls握手. )
比如跟gae搞h2这是xx的做法. 但是xx不给浏览器发送的流量正确转发alpn tls标签. 我猜的.
apln可能有时候会变 比如有些工具只有h1的alpn 所以我意思就是原封不动的转发下alpn 不是硬着写
我也不知道这样能不能用 说不定能用呢!!! 

我看了下privoxy的握手. 是出去的握手.
看起来他就是转发了数据么. 就跟我不用代理一个样.
是客户端的tls标签里面有h2,http/1.1
然后服务器端返回的*server hello* 里面alpn只有h2 这个时候客户端就会决定用h2传输数据.
目前不清楚跟代理是否有要求 但是值得尝试. 只需要转发下发送出去和接收的alpn就ok了.
也许我该找个2012年的代理服务器试试看 因为那时候绝对没支持h2的东西

好了不用测了 privoxy不支持h2. 他意思应该就是说 是客户端到privoxy的连接是h2的 因为那样就仅仅只会用1个连接. 这个不支持 但是可以支持转发h2数据 走h1.1.
这个因为h2就是单连接啊 所以走1.1影响也不大. 

4.16. Does Privoxy support HTTP/2?
Privoxy currently doesn't parse HTTP/2 but applications can **tunnel HTTP/2 through Privoxy** if Privoxy is configured to allow CONNECT requests (default) which are also used for HTTPS.

Adding HTTP/2 support is on the TODO list but currently nobody is known to work on it.

所以h1.1代理转发h2数据 和原生h2体验是一样的!!

**Comment by SeaHOH** [comment]
你可以再看看 readme 最后几行字。。。

**Comment by k79e** [comment]
这个还是1.1 就是完整转发下tls握手的那个alpn标签.
不知道gae能不能做到这个....

**Comment by SeaHOH** [comment]
准确的讲，那不叫转发，而是**中继**。
何为中继？就是将信号通过某个介质传递到它原先无法到达的地方，它只是增加范围，而不改变信号。

免费 GAE 代理做不到这一点，它必须改变，这是谷歌限制的。

**Comment by k79e** [comment]
GAE各种用urlfetch 这东西看起来好弱啊!!!!!
难不成就是因为这样么!!!
但是我看官网说还可以用其他库?? 这什么意思啊. 难不成还能绕过urlfetch!!?
2019-09-16 15:39:26
看起来可能没救了...

**Comment by k79e** [comment]
不清楚go的urlfetch是否能使用标准http库里面的那个Transport里面的各个功能.
如果可以的话那就太好了 那个里面有配置alpn的!! 一定就是发起https连接的那个tls配置!!!!
他这要是不支持 那我看收费的也没救啊 收费的不是也走urlfetch么?? 啊!!? 难道收费的不走urlfetch!!?

> For control over proxies, **TLS configuration**, keep-alives, compression, and other settings, create a Transport:

**Comment by SeaHOH** [comment]
谷歌又不傻，注意我讲的是免费应用。

**Comment by k79e** [comment]
昨天网络抽了 有个评论没发出去.
写的是不知道urlfetch能否使用标准库里的一个参数.就是http的Transport

**Comment by SeaHOH** [comment]
这几天正准备更新服务端，在看文档，你也可以看看下面这篇，注意它的中文是翻译来的，个别意思表示有误，可以对照英文阅读。
https://cloud.google.com/appengine/docs/standard/go112/how-requests-are-handled

**Comment by k79e** [comment]
我的天哪可以使用标准库!!! 看起来有希望了!!

**Comment by SeaHOH** [comment]
能用，但是功能有限制，跟现在也差不太多。

**Comment by k79e** [comment]
对了 我今天发现了关于sni的问题 想请教下
就是一般google 的ip 这个软件发送的sni我看是乱七八糟的 但是也能用
但是有些ip 我测试他可以用 是个镜像 但是sni乱发 服务器反而不认了.

我配置google ip列表的sni为 @none之后 也没见他不发送 他还是发送乱七八糟的sni.
这怎么办呢??

2019-09-18 18:17:28
我发填*.appspot.com还不行 用随意一个appspot的子域名就可以了....
我发现那个ip 打不开这个域名www.appspot.com 但是其他ip可以.
但是那个打不开www的 他可以开appspot.com的子域名 比如123.appspot.com
奇怪了 那我能不能用这个奇怪ip的时候 不发送sni呢? 

2019-09-18 18:30:42
我晕 这原来是个透明代理.... 我说怎么能打开呢..

**Comment by k79e** [comment]
我有一个这样的场景 不知道这个软件能不能这样
我appid是绑定域名的. 这个域名是个cdn.
我能否改个规则 让他 暂时绕过cdn. 就是让他绑定的地址 恢复google的那个原始cname.
然后我再给这个cname重新绑定下ip 也就是指定这个cname的ip

可恶 我想什么呢 hosts本身就可以那样用... 真是胡思乱想呢...

**Comment by SeaHOH** [comment]
> 对了 我今天发现了关于sni的问题 想请教下
> 就是一般google 的ip 这个软件发送的sni我看是乱七八糟的 但是也能用
> 但是有些ip 我测试他可以用 是个镜像 但是sni乱发 服务器反而不认了.
> 
> 我配置google ip列表的sni为 @none之后 也没见他不发送 他还是发送乱七八糟的sni.
> 这怎么办呢??

看主配置 [gae/servername]，这个是专为设置 GAE IP 的 SNI 而设计。默认就是发送无效 SNI，不能设置成不发送，优点是可以设置多个 SNI。
自动规则伪造 `.appspot.com` 的 SNI，这个优先级更高，可以设置成不发送。

很久没试过了，`*.appspot.com` 这种 SNI，不一定能用。

还有主配置 [picker/servername] 是针对 GWS IP 的，必须与 [picker/comdomain] 配对使用。
从逻辑上来讲直接用它不是那么合适，主要是偷懒少写个 [gws/servername] 配置。

**Comment by SeaHOH** [comment]
对了，我写的关于自动代理动作说明的 wiki 你看没有？它应该有助于你深入了解其运作原理。

**Comment by k79e** [comment]
看了啊 不看我怎么用的给cdn加hosts. 那个太好用了. 我正在研究cdn有没有什么各地的镜像呢

对了 那个随机sni 他能不能设定成 发送的虽然显示的有sni标签 但是里面是空的那种?
不知道gae认不认这样的.

cf这个cdn太坑人了 怎么找都找不到不是任播的地址. 这藏的也太深了吧.

**Comment by SeaHOH** [comment]
> 对了 那个随机sni 他能不能设定成 发送的虽然显示的有sni标签 但是里面是空的那种?

没试过，不知道 pyOpenSSL 是否支持，但是这个没必要吧，特征太明显。

**Comment by k79e** [comment]
我也不太清楚到底是空白sni项敏感还是连域名都不是的随机字符更危险.

**Comment by hg123** [comment]
我想请问下 
我要把goproxy上能用的sni的ip用在gotox上，这些ip应该填在哪个位置，还有其他哪些地方需要设置呢？
我试了好长时间都没试出来，麻烦帮我解答一下。谢谢了。

**Comment by SeaHOH** [comment]
@hg123 你说的是 SNI 反代吧，请查看以前的帖子，实际上配置注释里面也有少许说明。

特别说明一下，我不确定现在是否还能通过 SNI 反代连接，也不打算测试，你最好先确定这种方法现在还能使用，比如用 goproxy 测试一下。

**Comment by hg123** [comment]
@SeaHOH 是的 就是sni反代 
我在goproxy上能用 但就是在gotox上用不了 注释也看了 但还是不得其法，还盼指教。
另外还想问一下 google 10.1日后 是不是goproxy就不能用了？

**Comment by SeaHOH** [comment]
> 请查看以前的帖子

> google 10.1日后 是不是goproxy就不能用了？

暂时还能用，只是如果不使用新版本就无法部署更新。

**Comment by hg123** [comment]
谢谢了 
那我再去翻翻之前的帖子

**Comment by k79e** [comment]
请问GWS是什么意思. 我很久没用这类软件了...
我只知道GVS好像是他们视频服务器的. GAE就是GAE嘛.. 那GWS是啥我怎么也想不到了.....
对了楼上的. SNI太好用了 哈哈. 我试过了.
SNI就把那个ip直接填GAEip里面就能用了.
sni就跟透明代理一个样 一个hosts就接入了. 所以sni根本不存在软件支持不支持的问题吧....
只要可以指定域名走特定ip就都可以用这类的.

**Comment by SeaHOH** [comment]
> 请问GWS是什么意思.

wiki 有介绍。

> sni就跟透明代理一个样 一个hosts就接入了. 所以sni根本不存在软件支持不支持的问题吧....

只是有可能封 `.appspot.com` 这个 SNI 名称，看起来现在还能用。

**Comment by k79e** [comment]
这样啊.. 就是web服务器..  我果然猜不到了..... XD

**Comment by Quantum2Bits** [comment]
gae绑定域名后gotox不能访问gmail，但zebra可以。不知是不是没有设置好。以前xx-net也可以访问gmail的。

**Comment by SeaHOH** [comment]
如无特殊需求，大部分谷歌系网站通过 GotoX 直连就好，无需使用 GAE 代理。

**Comment by k79e** [comment]
我有点疑问
有些ip他证书是自签的 或者是其他的证书(买来的 或者是免费的那种证书) 反正不是google的
但是不验证证书却可以访问其他host 
这样给gotox用了安全么? 用了下也没报错什么的 很正常的样子.

2019-10-04 14:01:04
原来是没设置好 弄好了之后会报错. 这种事情想想都不安全 :D

**Comment by SeaHOH** [comment]
CA 证书是信任谷歌和火狐的证书库，还有 GotoX 自己生成的那个，其余都会报错。
以上只是针对 GotoX 本身直接请求的 TLS 连接，转发的连接不包含在内，需要客户端 (个别程序和脚本可能不会验证) 自行解决。
